pkg/event/processor.go

```diff
@@ -142,13 +148,26 @@ func (p *BatchEventProcessor) Start(exeCtx utils.ExecutionCtx) {
 
 // ProcessEvent processes the given impression event
 func (p *BatchEventProcessor) ProcessEvent(event UserEvent) {
-	p.Q.Add(event)
 
 	if p.Q.Size() >= p.MaxQueueSize {
-		go func() {
-			p.FlushEvents()
-		}()
+		pLogger.Warning("MaxQueueSize has been met. Discarding event")
+		return
 	}
+
+	p.Q.Add(event)
+
+	if p.Q.Size() >= p.BatchSize {
+		if processing, ok := p.processing.Get().(bool); ok && !processing {
```

Comment:
Perhaps you can use a [semaphore](https://godoc.org/golang.org/x/sync/semaphore).
```go
if p.sem.TryAquire(1) {
    go func() {
        p.FlushEvents()
        p.sem.Release(1)
    }()
}
```

ref: https://medium.com/@deckarep/gos-extended-concurrency-semaphores-part-1-5eeabfa351ce